numba/tracing.py

```diff
@@ -0,0 +1,207 @@
+
+    *args: categories in which to trace this function. Example usage:
+
+    @trace
+    def function(...):...
+
+    @trace('mycategory')
+    def function(...):...
+
+
+    """
+
+    recursive = kwds.get('recursive', False)
+    def decorator(func):
+
+        spec = None
+        logger = logging.getLogger('root')
+
+        def wrapper(*args, **kwds):
+            if not logger.isEnabledFor(logging.DEBUG) or tls.tracing:
+                return func(*args, **kwds)
+
+            fname, ftype = find_function_info(func, spec, args)
+            
+            try:
+                tls.tracing = True
+                enter, leave = create_events(fname, spec, args, kwds)
+
+                try:
+                    logger.debug(''.join(enter))
+                    tls.indent += 1
+                    try:
+                        try:
+                            tls.tracing = False
+                            result = func(*args, **kwds)
+                        finally:
+                            tls.tracing = True
+                    except:
+                        type, value, traceback = sys.exc_info()
+                        leave.append(' => exception thrown\n\traise ')
+                        mname = type.__module__
+                        if mname != '__main__':
+                            leave.append(mname)
+                            leave.append('.')
+                        leave.append(type.__name__)
+                        if value.args:
+                            leave.append('(')
+                            leave.append(', '.join(chop(v) for v in value.args))
+                            leave.append(')')
+                        else:
+                            leave.append('()')
+                        raise
+                    else:
+                        if result is not None:
+                            leave.append(' -> ')
+                            leave.append(chop(result))
+                finally:
+                    tls.indent -= 1
+                    logger.debug(''.join(leave))
+            finally:
+                tls.tracing = False
+            return result
+        # wrapper end
+
+        result = None
+        rewrap = lambda x: x
+        # Unwrap already wrapped functions
+        # (to be rewrapped again later)
+        if type(func) == classmethod:
+            rewrap = type(func)
+            # Note: 'func.__func__' only works in Python 3
+            func = func.__get__(True).__func__
+        elif type(func) == staticmethod:
+            rewrap = type(func)
+            # Note: 'func.__func__' only works in Python 3
+            func = func.__get__(True)
+        elif type(func) == property:
+            raise NotImplementedError
+
+        spec = inspect.getargspec(func)
+        return rewrap(wraps(func)(wrapper))
+    
+    logging.basicConfig(level = logging.DEBUG)
+
+    arg0 = len(args) and args[0] or None
+    # not supported yet...
+    if recursive:
```

Comment:
It will allow to call `trace(obj, recursive=True)` where 'obj' is a class instance or module whose attributes should be traced, too, i.e. where the decorator should be applied recursively.